ARMmbed/mbed-os
# CMake: Add CMake to mbed-psa greentea tests

### Summary of changes 


- Changed ```mbed_greentea_add_test``` macro to require TEST_SOURCES for all files, removing assumption of main.cpp
- Refactored all pre-existing greentea tests to add main.cpp using TEST_SOURCES
- Added a CMakeLists.txt for each test suite in Mbed OS PSA. 
- Moved /pal and /val directories into a separate /test_abstraction_layer directory and combined them into one CMake target, which is linked to by the test suites that require them.

Fixes #14680

#### Impact of changes 



None.

#### Migration actions required 

None.

### Documentation 


None. 

----------------------------------------------------------------------------------------------------------------
### Pull request type 


    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 


    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR

Manual testing for pre-existing greentea tests:

All pre-existing tests built successfully using below commands, or fail as expected due to a [NOT_SUPPORTED] error.
    
Manual testing for PSA tests:
Successfully able to build the test suite with below manual steps:
- ```mbedtools configure -t <toolchain> -m <target> --mbed-os-path <path-to-mbed-os> --app-config TESTS/configs/experimental.json``` (requires [this PR](https://github.com/ARMmbed/mbed-tools/pull/292))
- ```cd cmake_build/<target>/develop/<toolchain>/ && cmake ../../../.. -G Ninja -DMBED_TEST_LINK_LIBRARIES=mbed-os && cmake --build .```
- ```mbedhtrun -f <path/to/image> -p <port> -d <mount>```

All built successfully with both GCC_ARM and ARMC6, and run successfully on target K66F, except:
- entropy_inject: can only run with MBEDTLS_ENTROPY_NV_SEED defined; build fails with ```#error [NOT_SUPPORTED] PSA entropy injection tests can run only with MBEDTLS_ENTROPY_NV_SEED enabled.``` (Build successful if the macro is defined, but test crashes when run on K66F)
- compliance_its/test_s003: can only run with PS_ALLOW_ENTIRE_STORAGE_FILL defined; build fails with ```#error [NOT_SUPPORTED] Test is too long for CI, thus always fails on timeout.``` (Build successful if the macro is defined, and test runs to {{result;success}})
----------------------------------------------------------------------------------------------------------------
### Reviewers 



@Patater @LDong-Arm

----------------------------------------------------------------------------------------------------------------
